### DISCUSSION TITLE
Spawning a thread and call a function from lua

### DISCUSSION THREAD

**Discussion by ghosty2004** [mlua-rs/mlua#406] [Q&A] [answered]
Hi, I want to know if it's possible to export a `spawn_thread` function from a module which have a `LuaFunction` type argument and then call that function inside a rust thread?

I tried something like this:

```rust
use mlua::prelude::*;

fn spawn_thread(_: &Lua, func: LuaFunction) -> LuaResult<()> {
    println!("Spawning thread...");

    std::thread::spawn(move || {
        println!("Thread spawned");
        let _ = func.call(()).unwrap();
        println!("Thread finished");
    });

    Ok(())
}

#[mlua::lua_module]
fn my_module(lua: &Lua) -> LuaResult<LuaTable> {
    let exports = lua.create_table()?;
    exports.set("spawn_thread", lua.create_function(spawn_thread)?)?;
    Ok(exports)
}
```

but I get the following error from the compiler:
```
`*mut lua_State` cannot be shared between threads safely
within `mlua::lua::LuaInner`, the trait `Sync` is not implemented for `*mut lua_State`, which is required by `{closure@src/lib.rs:6:24: 6:31}: Send`
required for `Arc<mlua::lua::LuaInner>` to implement `Sync`
required for `&mlua::Lua` to implement `Send`
```

I'm using the following features: `lua53`, `module`, `send`

Any example about how can I achive multi thread like this will be appreciated.

**Comment by khvzak** [comment, marked as answer]
Lua (and mlua) is strictly `!Sync`, you cannot do this.
To achieve true parallelism you can spawn another Lua interpreter to execute function (passed as source code) then serialize results and pass back to main VM.

**Reply by cheesycod** [reply]
@khvzak can u send an example pls, I have this same issue on tokio
